walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -56,4 +54,85 @@
 
     return validTimes;
   }
+
+  /** 
+    * Returns a collection of times that work for an optimal number of optional 
+    * people, as well as all the mandatory people.  
+    */
+  private Collection<TimeRange> getOptimalSubsetTimes(
+    Set<String> mandatoryAttendees, Set<String> optionalAttendees, 
+    Collection<Event> events, MeetingRequest request) {
+    
+    Set<Set<String>> subsets = Sets.powerSet(optionalAttendees);
+    Set<Set<String>> infeasibleGroups = new HashSet<Set<String>>();
+    Collection<TimeRange> optimalListOfTimes = new ArrayList<TimeRange>();
+
+    for (int i = 1; i < optionalAttendees.size(); i++) {
+      int currentSizeOfSubsetsToCheck = i;
+      Set<Set<String>> subsetsToCheck = Sets.filter(subsets, subset -> 
+        subset.size() == currentSizeOfSubsetsToCheck && 
+          !containsInfeasibleGroup(subset, infeasibleGroups));
+      if (subsetsToCheck.isEmpty()) {
+        // all subsets larger than this will be infeasible 
+        break;
+      }
+
+      for (Set<String> subset : subsetsToCheck) {
```

Comment:
I tried to replace this for-each loop with a Stream.reduce() function that would produce the value that optimalListOfTimes should hold after iterating through the list of subsets, but because the accumulator was Collection<TimeRage> and the elements were Set<String>, it wouldn't compile, and I was having trouble finding a Combiner function that would make that work. 

I also tried using forEach(), which wouldn't compile because I was reassigning optimalListOfTimes. 

Is it alright that this is a for-each loop? I'm very open to ideas to make this more readable/optimized.

Reply:
LGTM